incubator/hnc/controllers/object_controller.go

```diff
@@ -102,63 +104,96 @@ func (r *ObjectReconciler) ReconcileNamespace(ctx context.Context, log logr.Logg
-	// Make sure this object is actually supposed to exist. If not, delete it. This should
-	// trigger one more reconciliation cycle and we'll get the children then.
-	if deleted, err := r.deleteObsolete(ctx, log, inst); deleted || err != nil {
+	// Make sure this object is correct and supposed to propagate. If not, delete it or mark it as
+	// modified. This should trigger one more reconciliation cycle and we'll get the children then.
+	if success, err := r.ensureCorrectAncestry(ctx, log, inst); !success || err != nil {
 		return err
 	}
 
 	// Make sure this gets propagated to any children that exist
 	return r.propagate(ctx, log, inst)
 }
 
-// deleteObsolete checks to see if the given object should not exist any more. It returns true if
-// it deleted the object and false otherwise.
-func (r *ObjectReconciler) deleteObsolete(ctx context.Context, log logr.Logger, inst *unstructured.Unstructured) (bool, error) {
-	// Find what we expect the source to be. If there's no source, *this*
-	// is the source and there's nothing to do.
+// getSourceNS gets source namespace if it's set in "labelInheritedFrom" label.
+// It returns an empty string if this instance is the source
+func getSourceNS(inst *unstructured.Unstructured) string {
 	labels := inst.GetLabels()
 	if labels == nil {
 		// this cannot be a copy
-		return false, nil
+		return ""
 	}
 	inheritedFrom, _ := labels[labelInheritedFrom]
-	if inheritedFrom == "" {
-		// this isn't a copy
-		return false, nil
+	return inheritedFrom
+}
+
+// getSourceInst gets source instance of this one. It returns nil if this instance isn't a copy.
+func (r *ObjectReconciler) getSourceInst(ctx context.Context, log logr.Logger, inst *unstructured.Unstructured) (bool, *unstructured.Unstructured, error) {
+	srcNS := getSourceNS(inst)
+	if srcNS == "" {
+		// this cannot be a copy
+		return false, nil, nil
 	}
 
-	// Read the source. If we find it, and the namespace is still an ancestor, no problem.
-	srcNm := types.NamespacedName{Namespace: inheritedFrom, Name: inst.GetName()}
+	// Get the source instance from source namespace
+	srcNm := types.NamespacedName{Namespace: srcNS, Name: inst.GetName()}
 	src := &unstructured.Unstructured{}
 	src.SetGroupVersionKind(inst.GroupVersionKind())
 	err := r.Get(ctx, srcNm, src)
-	if err == nil {
-		// The object exists. Is it still in an ancestor namespace?
-		if yes, err := r.isAncestor(inst.GetNamespace(), inheritedFrom); err != nil {
-			// This can fail if the names are bad. For now, take no action (ie don't
-			// delete). TODO: revisit this.
-			log.Error(err, "deleteObsolete")
-			return false, err
-		} else if yes {
-			// We found the object in an ancestor namespace; no need to delete it.
-			return false, nil
+	if err != nil {
+		if errors.IsNotFound(err) {
+			log.Info("The source object no longer exists", "inheritedFrom", srcNS)
+			return true, nil, err
```

Comment:
Don't return `err`, just return `nil` here because this isn't really an error to the caller

Reply:
done